core/src/main/scala/kafka/admin/ConsumerGroupCommand.scala

```diff
@@ -71,64 +72,122 @@ object ConsumerGroupCommand {
       }
     } catch {
       case e: Throwable =>
-        println("Error while executing consumer group command " + e.getMessage)
-        println(Utils.stackTrace(e))
+        consumerGroupOutputWriter.printError("Executing consumer group command failed due to " + e.getMessage, Some(e))
     } finally {
       consumerGroupService.close()
     }
   }
 
+  case class PartitionAssignmentState(group: String, topic: Option[String], partition: Option[Int], offset: Option[Long],
+                                      lag: Option[Long], memberId: Option[String], memberHost: Option[String],
+                                      clientId: Option[String], logEndOffset: Option[Long])
+
+  trait OutputWriter {
+    def printError(msg: String, e: Option[Throwable] = None)
+    def printConfirmation(msg: String)
+    def printAssignment(groupAssignment: Array[PartitionAssignmentState])
+  }
+
+  class ConsumerGroupOutputWriter extends OutputWriter with Logging {
+
+    def printError(msg: String): Unit = println("Error: " + msg)
+
+    def printError(msg: String, e: Option[Throwable] = None): Unit = {
+      println("Error: " + msg)
+      if (e.isDefined)
+        debug(Utils.stackTrace(e.get))
+    }
+
+    def printConfirmation(msg: String): Unit = println(msg)
+
+    def printAssignment(groupAssignment: Array[PartitionAssignmentState]): Unit = {
+      println("%-30s %-30s %-10s %-15s %-15s %-10s %-50s %-30s %s".format("GROUP", "TOPIC", "PARTITION", "CURRENT-OFFSET", "LOG-END-OFFSET", "LAG", "MEMBER-ID", "MEMBER-HOST", "CLIENT-ID"))
+      groupAssignment.foreach { memberAssignment =>
+        println("%-30s %-30s %-10s %-15s %-15s %-10s %-50s %-30s %s".format(
+          memberAssignment.group, memberAssignment.topic.getOrElse(""), memberAssignment.partition.getOrElse(""),
+          memberAssignment.offset.getOrElse(""), memberAssignment.logEndOffset.getOrElse(""), memberAssignment.lag.getOrElse(""),
+          memberAssignment.memberId.getOrElse("none"), memberAssignment.memberHost.getOrElse(""), memberAssignment.clientId.getOrElse("")))
+      }
+    }
+  }
+
   sealed trait ConsumerGroupService {
 
     def list(): Unit
 
-    def describe() {
-      describeGroup(opts.options.valueOf(opts.groupOpt))
+    def describe(): Option[Array[PartitionAssignmentState]] = {
```

Comment:
That's kind of unfortunate. So here's another idea (feel free to dismiss it if it doesn't make sense). Maybe having `describe()` return the assignment is actually heading in the right direction. What if we get rid of `OutputWriter` and move the printing logic into the main method. Then the `ConsumerGroupService` becomes more functional (and testable). For `list()`, you can have it return the list of groups instead of printing them. The trickier one seems to be `delete()`. Maybe you can leave it as it is.